stardist segmentation example

AnnaChristina opened the pull request:
example notebook for stardist segmentation

hspitzer commented on docs/source/external_tutorials/tutorial_stardist.ipynb:
```diff
@@ -0,0 +1,517 @@
+   "source": [
+    "# Nuclei segmentation using stardist\n",
+    "\n",
+    "This tutorial shows how we can easily use an external segmentation packages in the *Squidpy* framework. Tp use the stardist package, pip-install stardist with `pip install stardist` and additionally `pip install tensorflow` as this is a requirement of the stardist package. For details on the segmentation model, see: Schmidt U., et al. *Cell Detection with Star-Convex Polygons.* MICCAI 2018 P ...
```
```suggestion
    "In this tutorial we show how we can use the `stardist` in `squidpy.im.segment` for nuclei segmentation. To use the stardist package, pip-install stardist with `pip install stardist` and additionally `pip install tensorflow` as this is a requirement of the stardist package. For details on the segmentation model, see: Schmidt U., et al. *Cell Detection with Star-Convex Polygons.* MICCAI 2018 Proceedings, Part II (265-273). doi: 10.1007/978-3-030-00934-2_30 or https://github.com/stardist/stardist for the GitHub repository.\n",
```

hspitzer replied:
could you also add 1 line describing stardist and linking to the paper there? Similar to the tangram tutorial: https://squidpy.readthedocs.io/en/latest/external_tutorials/tutorial_tangram.html
**Tangram** (<cite data-cite="tangram">Biancalani et al. (2020)</cite>, [code](https://github.com/broadinstitute/Tangram)). The reference goes in docs/source/references.bib I think

hspitzer commented on docs/source/external_tutorials/tutorial_stardist.ipynb:
```diff
@@ -0,0 +1,517 @@
+    "\n",
+    "fig, axes = plt.subplots(1, 3, figsize=(10, 20))\n",
+    "for i, ax in enumerate(axes):\n",
+    "    crop.show(\"image\", channel=i, ax=ax)"
```
we can use `crop.show(channelwise=True)` here.

hspitzer reviewed (commented):
Looks great, thanks for also including an option to have a smaller threshold in the H&E image!

Can you also do the following (copying parts of Michals comments on another PR here):
1. add the new tutorial to `docs/source/tutorial.rst`
2. run `tox -e check-docs` (will alert you to typos)
3. run `tox -e shallow-docs` to see if it renders as expected
6. add a thumbnail image? e.g. for tangram, the NB cell with the image has additional metadata
```
   "metadata": {
    "nbsphinx-thumbnail": {
     "output-index": 1,
     "tooltip": "Integrate the Image Container with Tangram for cell-type deconvolution."
    }
   },
```

AnnaChristina commented:
@hspitzer adjusted according to your comments and passing tox checks.

giovp commented:
thanks @AnnaChristina ! 
- Can you also add a `stardist_env.yml` file with your conda env?
- squidpy needs to be installed with pip so either re-create env or manually change the yml file with the latest squidpy (no pin version), so no worry if version don't match here (it will eventually)
- just have minimal reproducible environment
- you can add the env file in the main folder, we can move it around later
- check out some notes that possibly apply here, froma recent similar notebook: https://github.com/theislab/squidpy_notebooks/pull/63#pullrequestreview-661239846
thanks a lot again 🙏

AnnaChristina commented:
according to https://github.com/theislab/squidpy_notebooks/pull/63#pullrequestreview-661239846
> 1. add the new tutorial to `docs/source/tutorial.rst`

- [x] added 

> 2. run `tox -e check-docs` (there are some typos)

- [x] ran 

> 3. fix the heading levels? (run `tox -e shallow-docs` to see if it renders as expected)

- [x] renders as expected

> 4. remove the `load_ext ...` cell and the last empty cell
> 5. remove the `.DS_Store` files from commit + maybe add to .gitignore

- [x] not applying

> 6. add a thumbnail image? e.g. for tangram, the NB cell with the image has additional metadata
> 
> ```
>    "metadata": {
>     "nbsphinx-thumbnail": {
>      "output-index": 1,
>      "tooltip": "Integrate the Image Container with Tangram for cell-type deconvolution."
>     }
>    },
> ```
> 
> (not sure whether this will work for cells where you load the image externally, but hope so).

- [x] added last image as nb-thumbnail


and according to @giovp comments:
- [x] added stardist_environment.yml

giovp commented:
to me it looks great! if @hspitzer we can merge!

hspitzer commented on stardist_environment.yml:
```diff
@@ -0,0 +1,12 @@
+  - pip=21.1.1
+  - python=3.8
+  - pip:
+    - squidpy==1.0.1.dev17+g6ba1291
```
@giovp should we specify the version here? Maybe just Squidpy is better?

giovp replied:
Yeah agree, we can remove the pin, also for python and pip (were probably set by `conda env export`)

hspitzer reviewed (approved):
Thanks, @AnnaChristina looks good to merge for me! 
I just added an explicit mention of the environment.yml and changed the heading levels for nicer appearance in the docs. 
Only comment is about the squidpy version in the yml, I'm not sure what is best, but I think I'd prefer to not pin a specific version, because later versions should work as well. Is it possible to do `>=xxx` for pip-installable packages as well?

giovp commented on stardist_environment.yml:
```diff
@@ -0,0 +1,12 @@
+  - pip=21.1.1
+  - python=3.8
+  - pip:
+    - squidpy==1.0.1.dev17+g6ba1291
```
```suggestion
    - squidpy
```

giovp commented on stardist_environment.yml:
```diff
@@ -0,0 +1,12 @@
+channels:
+  - conda-forge
+dependencies:
+  - pip=21.1.1
```
```suggestion
  - pip
```

giovp commented on stardist_environment.yml:
```diff
@@ -0,0 +1,12 @@
+  - conda-forge
+dependencies:
+  - pip=21.1.1
+  - python=3.8
```
```suggestion
  - python>=3.8
```

hspitzer commented:
last comment, @AnnaChristina I noticed that you have some code for smoothing the h&e image, but don't actually use "image_smooth" in the segmentation. Could you either remove the smoothing, or change stardist to use the smoothed image? Imo its fine to remove the smoothing here, the results look good even without it!

AnnaChristina commented:
> last comment, @AnnaChristina I noticed that you have some code for smoothing the h&e image, but don't actually use "image_smooth" in the segmentation. Could you either remove the smoothing, or change stardist to use the smoothed image? Imo its fine to remove the smoothing here, the results look good even without it!

@hspitzer done! I also changed the images to full image plus segmentation for H&E examples as StarDist anyway works on all three channels!
